docs/source/generics.rst

```diff
@@ -0,0 +1,984 @@
+Generics
+========
+
+You may have seen type hints like ``list[str]`` or ``dict[str, int]`` in Python
+code. These types are interesting in that they are parametrised by other types!
+A ``list[str]`` isn't just a list, it's a list of strings. Types with type
+parameters like this are called *generic types*.
+
+You can define your own generic classes that take type parameters, similar to
+built-in types such as ``list[X]``. Note that such user-defined generics are a
+moderately advanced feature and you can get far without ever using them.
+
+.. _generic-classes:
+
+Defining generic classes
+************************
+
+Here is a very simple generic class that represents a stack:
+
+.. code-block:: python
+
+   from typing import TypeVar, Generic
+
+   T = TypeVar('T')
+
+   class Stack(Generic[T]):
+       def __init__(self) -> None:
+           # Create an empty list with items of type T
+           self.items: list[T] = []
+
+       def push(self, item: T) -> None:
+           self.items.append(item)
+
+       def pop(self) -> T:
+           return self.items.pop()
+
+       def empty(self) -> bool:
+           return not self.items
+
+The ``Stack`` class can be used to represent a stack of any type:
+``Stack[int]``, ``Stack[tuple[int, str]]``, etc.
+
+Using ``Stack`` is similar to built-in container types, like ``list``:
+
+.. code-block:: python
+
+   # Construct an empty Stack[int] instance
+   stack = Stack[int]()
+   stack.push(2)
+   stack.pop() + 1
+   stack.push('x')  # error: Argument 1 to "push" of "Stack" has incompatible type "str"; expected "int"
+
+When creating instances of generic classes, the type parameter can usually be
+inferred. In cases where you explicitly specify the type parameter, the
+construction of the instance will be type checked correspondingly.
+
+.. code-block:: python
+
+   class Box(Generic[T]):
+       def __init__(self, content: T) -> None:
+           self.content = content
+
+   Box(1)       # OK, inferred type is Box[int]
+   Box[int](1)  # Also OK
+   Box[int]('some string')  # error: Argument 1 to "Box" has incompatible type "str"; expected "int"
```

Comment:
Is it worth elaborating what "compatibility" means here?

My understanding is that class instantiations require argument types to be consistent subtypes of parameter types. So, here `str` is not a consistent subtype of `int`, thus the error. PEP-483 [calls](https://peps.python.org/pep-0483/#summary-of-gradual-typing) consistent subtyping just "consistency".

Reply:
Like Alex says below, I think okay to have slightly lax terminology. This is meant to be approachable documentation, not necessarily a spec, and is pretty intuitive that str is not compatible with int